packages/ic-http-certification/src/http/http_request.rs

```diff
@@ -1,39 +1,766 @@
 use crate::{HeaderField, HttpCertificationError, HttpCertificationResult};
 use candid::{CandidType, Deserialize};
 use http::Uri;
+use std::borrow::Cow;
 
-/// A Candid-encodable representation of an HTTP request.
-/// This struct is used by canisters that implement the HTTP interface required by the HTTP Gateway Protocol.
+/// A Candid-encodable representation of an HTTP request. This struct is used by
+/// the `http_request` method of the HTTP Gateway Protocol's Candid interface.
+///
+/// # Examples
+///
+/// ```
+/// use ic_http_certification::HttpRequest;
+///
+/// let request = HttpRequest::builder()
+///     .with_method("GET")
+///     .with_url("/")
+///     .with_headers(vec![("X-Custom-Foo".into(), "Bar".into())])
+///     .with_body(&[1, 2, 3])
+///     .with_certificate_version(2)
+///     .build();
+///
+/// assert_eq!(request.method(), "GET");
+/// assert_eq!(request.url(), "/");
+/// assert_eq!(request.headers(), &[("X-Custom-Foo".into(), "Bar".into())]);
+/// assert_eq!(request.body(), &[1, 2, 3]);
+/// assert_eq!(request.certificate_version(), Some(2));
+/// ```
 #[derive(Clone, Debug, CandidType, Deserialize, PartialEq, Eq)]
-pub struct HttpRequest {
+pub struct HttpRequest<'a> {
     /// HTTP request method.
-    pub method: String,
-    /// Request URL.
-    pub url: String,
+    method: Cow<'a, str>,
+
+    /// HTTP request URL.
+    url: Cow<'a, str>,
```

Comment:
As discussed, using `Cow` for these fields seems like an overkill, that complicates the code unnecessarily.  If that's fine for you, maybe revert to `String`?

Reply:
While making the change, I realized that this suggestion will make the library code slightly cleaner, but the consuming code now needs to call `to_string()` for any string references.

With Cow:
```rust
let request = HttpRequest::patch("/").build();
```

versus:
```rust
let request = HttpRequest::patch("/".to_string()).build();
```

Using `Cow` for efficiency is overkill, but it also has the benefit of making the library interface more flexible and allowing consumers to pass either references, or owned values, depending on what is most convenient for them.

What do you think?

Reply:
Not sure I understand: we would have `HttpRequest::patch(url: &str)`, and then one would still make a call like `HttpRequest::patch("/").build()` (while we'd do the conversion `url.to_string()` internally in `patch()`), no?  Or am I missing something?

Reply:
We agreed to use owned `String` values, but expose `str` reference and use `to_string()` internally.